sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs

```diff
@@ -0,0 +1,231 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4057";
+        private const string MessageFormat = "Set the locale for this '{0}'.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<KnownType> invalidTypes = new HashSet<KnownType>
+        {
+            KnownType.System_Data_DataTable,
+            KnownType.System_Data_DataSet
+        };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(
+                compilationStartContext =>
+                {
+                    var symbolsWhereTypeIsCreated =
+                        new ConcurrentBag<SyntaxNodeWithSymbol<ObjectCreationExpressionSyntax, ISymbol>>();
+                    var symbolsWhereLocaleIsSet = new ConcurrentBag<ISymbol>();
+
+                    compilationStartContext.RegisterSyntaxNodeActionInNonGenerated(
+                        c =>
+                        {
+                            var objectCreation = (ObjectCreationExpressionSyntax)c.Node;
+                            var objectType = c.SemanticModel.GetSymbolInfo(objectCreation.Type).Symbol as ITypeSymbol;
+
+                            if (objectType == null ||
+                                !objectType.IsAny(invalidTypes))
+                            {
+                                return;
+                            }
+
+                            var variableSyntax = GetAssignedToSyntax(objectCreation);
+                            if (variableSyntax == null)
+                            {
+                                return;
+                            }
+
+                            var variableSymbol = variableSyntax is IdentifierNameSyntax
+                                ? c.SemanticModel.GetSymbolInfo(variableSyntax).Symbol
+                                : c.SemanticModel.GetDeclaredSymbol(variableSyntax);
```

Comment:
I was curious if this is a mistake, because we talked about it and I was left with the impression that only one is needed in this case. But if it isn't, I am fine with the solution.